clean_dockerhub.py

```diff
@@ -0,0 +1,277 @@
+    and the method returns True. If no such exception is raised, the method returns False.
+    """
+
+    # Higher values will run deletion faster, but have higher risk of getting a 429,
+    # see https://docs.docker.com/docker-hub/usage/#abuse-rate-limit
+    # Tests have shown that max deletion rate without running into rate limits is ~650 Tags/min.
+    sem = asyncio.Semaphore(2)
+
+    retry = False
+    conn = aiohttp.TCPConnector(limit_per_host=10)
+    async with aiohttp.ClientSession(connector=conn) as session:
+        try:
+            tasks = [
+                asyncio.create_task(
+                    delete_tag(session, namespace, repo, tag, token, sem)
+                )
+                for tag in tags_to_delete
+            ]
+
+            status_counter_map: dict[int, int] = {}
+            retry = False
+
+            for delete_coro in tqdm(
+                asyncio.as_completed(tasks),
+                total=len(tasks),
+                desc="Deleting old tags",
+            ):
+                try:
+                    status = await delete_coro
+                    status_counter_map[status] = status_counter_map.get(status, 0) + 1
+                except TooManyRequestsError as e:
+                    # upon 429, cancel all pending tasks
+                    retry = True
+                    for t in tasks:
+                        if not t.done():
+                            t.cancel()
+                    break
+
+        finally:
+            # ensure all tasks are concluded to suppress warnings
+            await asyncio.gather(*tasks, return_exceptions=True)
+
+        print(
+            f"Deletion run completed. \n Status counters: \n {status_counter_map}\n Retry: {retry}",
+        )
+    return retry
+
+
+async def clean_dockerhub(
+    docker_repository: str,
+    docker_username: str,
+    docker_password: str,
+    min_age_in_days: int,
+    max_number_of_pages: int,
+) -> None:
+    """
+    Cleans all tags in a Docker repository which are older than `min_age_in_days` days.
+    1. Gets the auth token from Docker Hub.
+    2. Fetches tags from Docker Hub (max. 10.000)
+    3. Deletes the tags in a loop until all tags are deleted, or only errors != 429 were raised.
+    """
+
+    namespace, repo = docker_repository.split("/")
+
+    # 1) Authenticate
+    async with aiohttp.ClientSession() as session:
+
+        token = await get_jwt_token(session, docker_username, docker_password)
+        print("Authenticated successfully.")
+
+    # 1) Fetch old tags
+    old_tags = await get_old_tags(
+        namespace, repo, token, min_age_in_days, max_number_of_pages
+    )
+    if not old_tags:
+        print("No tags to delete. Exiting.")
+        return
+
+    # 2) Delete old tags. Repeat until we don't run into a rate limit
+    while True:
+        tags_to_delete = [tag for tag in old_tags if not tag.deleted]
+        retry = await delete_tags(namespace, repo, token, tags_to_delete)
```

Comment:
I think this answers my comment above. 
So the semantics of the boolean is "needs_retry". :slightly_smiling_face: 
I updated my comment and suggestion above.